### PR TITLE
Support for computing arbitrary order derivatives of 2-body integrals with Engine

### PR DESCRIPTION
**Pull request by adabbott**
This PR adds support for 2-body integral Engines which compute 3rd and 4th derivatives. I have tested this current implementation up to d functions for standard 4-center `operator::coulomb` integrals only, and it is working as intended. Code has been added for 2-center and 3-center coulomb integrals as well, but not tested as my current libint configuration does not support their derivatives, and I do not have a reference implementation to test against. Hence, that part of the code is commented out for now.

@evaleev let me know if you want me to change anything. In particular,
* The lookup-arrays `mapDerivIndexN_xxxx` are very large and unsightly at higher orders, do you think they should instead be imported from another header file, and loaded in when the Engine object is instantiated with a particular `deriv_order`? Suggestions welcome
* I generated those new `mapDerivIndexN_xxxx` arrays with a short NumPy script, and this can be added to the PR if desired, or shared by other means.
* I am not sure how to implement the relevant unit tests

### CODE REVIEW

**Comment by evaleev** [comment]
@adabbott thank you, this looks like a great step towards the long-time goal!

Qs/comments:
- how hard would it be to port the logic of the script generating the maps to a C++ function? Ideally we could just do this for arbitrary orders ... the static maps were purely used to save time / indulge laziness.
- the map generation can be made purely static and triggered by libint2::initialize()
- ad hoc unit tests should go into `tests/unit/test-2body.cc`
- integration testing of libint is done via hartree-fock++, in the latter I would just compute the "core" contribution to the 2-body part of 3-rd/4-th deriv tensor and hardwire the reference results to hartree-fock++-validate.py

Thoughts?

**Comment by adabbott** [comment]
@evaleev Generating the maps with a C++ function sounds like a great idea. It shouldn't be too difficult to convert my python code to C++. I think I can write a universal function `generate_deriv_index_map` which given a deriv_order and BraKet type can output any particular mapDerivIndex* that currently appears in `engine.impl.h`.  Do you have a suggestion as to where I would put this function? Also, where should the map array be created? [Here](https://github.com/evaleev/libint/blob/3bf3a07b58650fe2ed4cd3dc6517d741562e1249/include/libint2/engine.impl.h#L682), perhaps? 

I'll look into adding tests once this is done.

**Comment by evaleev** [comment]
@adabbott great! I would make these tables managed by a dedicated singleton class whose entire purpose is to manage the initialization/lifetime and access from `Engine` ... see https://github.com/evaleev/libint/blob/master/include/libint2/solidharmonics.h#L82 for an example ... since `Engine` constructors can be called in general from multiple threads I would make the maps statically initialized up to the order that the library actually supports (this way you can avoid needing to lock the table of maps in case it needs to be extended to support derivatives of order higher than requested previously).

**Comment by adabbott** [comment]
@evaleev I think this is ready for another round of feedback. As this is my first substantial contribution to a C++ library, I am a bit out of my comfort zone, so nitpicking is very much welcomed. Some things that are worth noting:

1. Currently in `initialize.h` I have it calling `DerivMapGenerator::initialize()`, which computes and stores all of the mapDerivIndex arrays needed. This is working, but is this the correct place to put it?

2. Am I right in assuming that the definitions in `engine.impl.h` regarding LIBINT2_CENTER_DEPENDENT_MAX_AM_3eri... can be removed and instead just always set `int ket_lmax = hard_default_lmax_`?

3. For `DerivMapGenerator::instance` I originally wanted to pass a BraKet type to this function, but including `engine.h` to give access to the enum class causes some strange errors. Currently I am instead passing an integer `ncenters` to distinguish between `BraKet::xx_xx` and `BraKet::xs_xx`, but this is probably not general enough as it does not distinguish between
`BraKet::xs_xx` and `BraKet::xx_xs`, for example. I could change this function to accept a string, "xx_xx", "xs_xx", etc so that it is more easily extensible, but I'm open to better options. WDYT?

4. `DerivMapGenerator::generate_deriv_index_map` only generates the index mappings once when libint is initialized, for every BraKet type and deriv order up to LIBINT2_MAX_DERIV_ORDER. The overhead of doing this is minimal: generating all map arrays up to 6th-order derivatives for both Braket::xx_xx and BraKet::xs_xx takes about 300 ms and uses 4 kB on my machine.

File: include/libint2/engine.impl.h [hunk 1]
```diff
@@ -1687,6 +1696,9 @@ __libint2_engine_inline const Engine::target_ptr_vec& Engine::compute2(
 
       case BraKet::xx_xs: assert(false && "this braket is not supported"); break;
       case BraKet::xs_xx: {
+        // TODO, now that we have arbitrary-order deriv maps, this logic is no longer needed, right?
```
Comment (evaleev) [inline include/libint2/engine.impl.h:1699]: It seems (per https://github.com/evaleev/libint/blob/master/include/libint2/engine.impl.h#L600):
```
const int ket_lmax = (hard_default_lmax_ == std::numeric_limits<int>::max()) ? hard_lmax_ : hard_default_lmax_;
```

**Review by evaleev — CHANGES_REQUESTED** [review]
Thanks, looks great!

File: include/libint2/deriv_map.h [hunk 2]
```diff
@@ -0,0 +1,239 @@
...
+#include <libint2/util/cxxstd.h>
+#if LIBINT2_CPLUSPLUS_STD < 2011
+# error "The simple Libint API requires C++11 support"
+#endif
+
+#include <stdlib.h>
+#include <vector>
+#include <algorithm>
+// Initially tried this so I could look at BraKet type directly, rather than `ncenters`, but failed 
+//#include "engine.h"
```
[the remark this answers is not in the document]
Comment (adabbott) [reply include/libint2/deriv_map.h:34]: The problem was not clear to me. When building the first HF test in the test suite, there was about 10,000 lines (guessing, it filled my terminal history) of complaints of the nature "x has not been declared", "y is not a member of libint2", "invalid use of incomplete type z". I will try moving BraKet to its own dedicated header, that sounds like a better framework than what I am currently doing.

File: include/libint2/engine.impl.h [hunk 3]
```diff
@@ -666,6 +667,14 @@ __libint2_engine_inline void Engine::initialize(size_t max_nprim) {
 
   if (max_nprim != 0) primdata_.resize(std::pow(max_nprim, braket_rank()));
 
+  // Grab pointer to derivative index permutation map
+  if (deriv_order_ > 0) {
+    int ncenters;
+    if (braket_ == BraKet::xx_xx) ncenters = 4; 
+    if (braket_ == BraKet::xs_xx) ncenters = 3;
```
[the remark this answers is not in the document]
Comment (adabbott) [reply include/libint2/engine.impl.h]: I thought the two center case [does not need a permutation map](https://github.com/evaleev/libint/blob/3bf3a07b58650fe2ed4cd3dc6517d741562e1249/include/libint2/engine.impl.h#L1864)?  Also [here](https://github.com/evaleev/libint/blob/3bf3a07b58650fe2ed4cd3dc6517d741562e1249/include/libint2/engine.impl.h#L1949)

File: include/libint2/engine.impl.h (hunk 1 above)
Comment (adabbott) [reply include/libint2/engine.impl.h]: Done

**Comment by adabbott** [comment]
@evaleev I think this is ready for another review. All tests are passing locally when I configure `--with-max-am=2 --with-opt-am=0 --enable-1body=2 --enable-eri=2 --enable-eri3=2 --enable-eri2=2`, but for some reason the Travis build seems to be consistently failing when it gets to the pearl test scripts in `src/bin/test_eri/` ... any idea why?

**Comment by evaleev** [comment]
the failing CI job fails due to GH actions jitter, rest looks good.

**Pull request merged**